mso/resource_mso_schema.go

```diff
@@ -107,70 +146,111 @@ func resourceMSOSchemaUpdate(d *schema.ResourceData, m interface{}) error {
-			"path": "/displayName",
-			"value": "%s"
-		}
-	`, name)
+`, newTemplate, newTemplate)
 
-	templateNamePayload := fmt.Sprintf(`
-		{
-			"op": "replace",
-			"path": "/templates/%s/name",
-			"value": "%s"
-		}
-	`, oldTemplate, newTemplate)
+		jsonSchema, err := container.ParseJSON([]byte(schemaNamePayload))
+		jsonTemplate, err := container.ParseJSON([]byte(templateNamePayload))
+		jsonDispl, err := container.ParseJSON([]byte(tempDisplayNamePayload))
+		payloadCon := container.New()
 
-	tempDisplayNamePayload := fmt.Sprintf(`
-		{
-			"op": "replace",
-			"path": "/templates/%s/displayName",
-			"value": "%s"
+		payloadCon.Array()
+		err = payloadCon.ArrayAppend(jsonSchema.Data())
+		if err != nil {
+			return err
 		}
-	`, newTemplate, newTemplate)
-
-	jsonSchema, err := container.ParseJSON([]byte(schemaNamePayload))
-	jsonTemplate, err := container.ParseJSON([]byte(templateNamePayload))
-	jsonDispl, err := container.ParseJSON([]byte(tempDisplayNamePayload))
-	payloadCon := container.New()
-
-	payloadCon.Array()
-	err = payloadCon.ArrayAppend(jsonSchema.Data())
-	if err != nil {
-		return err
-	}
-	payloadCon.ArrayAppend(jsonTemplate.Data())
-	payloadCon.ArrayAppend(jsonDispl.Data())
-	path := fmt.Sprintf("api/v1/schemas/%s", d.Id())
+		payloadCon.ArrayAppend(jsonTemplate.Data())
+		payloadCon.ArrayAppend(jsonDispl.Data())
+		path := fmt.Sprintf("api/v1/schemas/%s", d.Id())
 
-	req, err := msoClient.MakeRestRequest("PATCH", path, payloadCon, true)
-	if err != nil {
-		return err
-	}
-	cont, _, err := msoClient.Do(req)
-	if err != nil {
-		return err
-	}
+		req, err := msoClient.MakeRestRequest("PATCH", path, payloadCon, true)
+		if err != nil {
+			return err
+		}
+		cont, _, err := msoClient.Do(req)
+		if err != nil {
+			return err
+		}
 
-	err = client.CheckForErrors(cont, "PATCH")
-	if err != nil {
-		return err
+		err = client.CheckForErrors(cont, "PATCH")
+		if err != nil {
+			return err
+		}
+	} else {
+		if d.HasChange("template") {
+			old_templates, new_templates := d.GetChange("template")
+			platform := msoClient.GetPlatform()
+
+			//In non-ND based MSOs, tenant cannnot be changed by PATCH API, so we delete it explicitly.
+			if platform != "nd" {
+				//Get all the new maps
+				getDifferenceNew := differenceInMaps(new_templates.(*schema.Set), old_templates.(*schema.Set))
+
+				// Get old maps that have a change
+				getDifferenceOld := differenceInMaps(old_templates.(*schema.Set), new_templates.(*schema.Set))
+
+				for _, valueMapOld := range getDifferenceOld {
+					valueOld := valueMapOld.(map[string]interface{})
+					for _, valueMapNew := range getDifferenceNew {
+						valueNew := valueMapNew.(map[string]interface{})
+
+						//Tenant ID has been changed. Delete Template.
+						if valueOld["name"] == valueNew["name"] && valueOld["display_name"] == valueNew["display_name"] {
+							deleteTemplate(valueOld["name"].(string), valueOld["display_name"].(string), valueOld["tenant_id"].(string), d, m)
+						}
+					}
```

Comment:
How do you take care of the templates that are removed from the main.tf are they in differenceOld? But they will not have a valueNew...

Reply:
MSO's PATCH API works well for the templates that are removed from the configuration file. It also works well when name of the template and display name of the template are changed. It only complains when we try to change tenant id of a particular template. Hence,  when the user wants to change the tenant id of a particular template, the above code deletes that particular template and re create it.

For NDO, PATCH does not require deleting a template explicitly when the tenant id is changed. Hence we bypass the deletion of template step for NDO.